tools/ssvm/main.cpp

```diff
@@ -1,41 +1,181 @@
 #include "common/value.h"
+#include "host/wasi/wasimodule.h"
+#include "po/argument_parser.h"
+#include "support/filesystem.h"
 #include "vm/configure.h"
 #include "vm/vm.h"
 
+#include <cstdlib>
 #include <iostream>
 
-int main(int Argc, char *Argv[]) {
-  if (Argc < 3) {
-    /// Arg0: ./ssvm
-    /// Arg1: wasm file
-    /// Arg2: invoke function name
-    /// Arg3...: inputs
-    std::cout << "Usage: ./ssvm wasm_file.wasm func_name [args...]"
-              << std::endl;
+int main(int Argc, const char *Argv[]) {
+  namespace PO = SSVM::PO;
+  using namespace std::literals;
+
+  std::ios::sync_with_stdio(false);
+  SSVM::Log::setErrorLoggingLevel();
+
+  PO::Option<std::string> WasmName(PO::Description("Wasm file"s),
+                                   PO::MetaVar("WASM_FILE"s));
+  PO::List<std::string> Args(PO::Description("Execution arguments"s),
+                             PO::MetaVar("ARG"s));
+
+  PO::Option<PO::Toggle> Reactor(PO::Description(
+      "Enable reactor mode. Reactor mode calls `_initialize` if exported."));
+
+  PO::List<std::string> Dir(
+      PO::Description(
+          "Binding directories into WASI virtual filesystem. Each directories "
+          "can specified as --dir `host_path:guest_path`, where `guest_path` "
+          "specifies the path that will correspond to `host_path` for calls "
+          "like `fopen` in the guest."s),
+      PO::MetaVar("PREOPEN_DIRS"s));
+
+  PO::List<std::string> Env(
+      PO::Description(
+          "Environ variables. Each variables can specified as --env `NAME=VALUE`."s),
+      PO::MetaVar("ENVS"s));
+
+  if (!PO::ArgumentParser()
+           .add_option(WasmName)
+           .add_option(Args)
+           .add_option("reactor", Reactor)
+           .add_option("dir", Dir)
+           .add_option("env", Env)
+           .parse(Argc, Argv)) {
     return 0;
   }
 
-  std::string InputPath(Argv[1]);
+  std::string InputPath = std::filesystem::absolute(WasmName.value()).string();
   SSVM::VM::Configure Conf;
+  Conf.addVMType(SSVM::VM::Configure::VMType::Wasi);
   SSVM::VM::VM VM(Conf);
 
-  /// Parameters and return values.
-  std::vector<SSVM::ValVariant> Params, Results;
-  uint32_t Err = 0;
+  SSVM::Host::WasiModule *WasiMod = dynamic_cast<SSVM::Host::WasiModule *>(
+      VM.getImportModule(SSVM::VM::Configure::VMType::Wasi));
 
-  for (int I = 3; I < Argc; I++) {
-    Params.push_back(static_cast<uint32_t>(std::stoul(Argv[I])));
-  }
-  if (auto Res = VM.runWasmFile(InputPath, Argv[2], Params)) {
-    Results = *Res;
-    for (auto &It : Results) {
-      std::cout << " Return value: " << std::get<uint32_t>(It) << std::endl;
+  WasiMod->getEnv().init(Dir.value(), WasmName.value(), Args.value(),
+                         Env.value());
+
+  if (!Reactor.value()) {
+    // command mode
+    if (auto Result = VM.runWasmFile(InputPath, "_start")) {
+      return WasiMod->getEnv().getExitCode();
+    } else {
+      std::cout << "Failed. Error code : "
```

Comment:
You should use `std::cerr` or our logging system APIs to report an error.

Reply:
Updated.